datalad/plugin/export_archive.py

```diff
@@ -0,0 +1,126 @@
+      Type of archive to generate.
+    compression : {'', 'gz', 'bz2')
+      Compression method to use. 'bz2' is not supported for ZIP archives.
+    on_failure : {'error', 'continue', 'ignore'}, optional
+      By default, any issue accessing a file in the dataset while adding
+      it to the TAR archive will result in an error and the plugin is
+      aborted. Setting this to 'continue' will issue warnings instead
+      of failing on error. The value 'ignore' will only inform about
+      problem at the 'debug' log level. The latter two can be helpful
+      when generating a TAR archive from a dataset where some file content
+      is not available locally.
+
+    """
+    import os
+    import tarfile
+    import zipfile
+    from mock import patch
+    from os.path import join as opj, dirname, normpath, isabs
+    from datalad.utils import file_basename
+    from datalad.support.annexrepo import AnnexRepo
+    from datalad.dochelpers import exc_str
+
+    import logging
+    lgr = logging.getLogger('datalad.plugin.export_archive')
+
+    repo = dataset.repo
+    committed_date = repo.get_committed_date()
+
+    # could be used later on to filter files by some criterion
+    def _filter_tarinfo(ti):
+        # Reset the date to match the one of the last commit, not from the
+        # filesystem since git doesn't track those at all
+        # TODO: use the date of the last commit when any particular
+        # file was changed -- would be the most kosher yoh thinks to the
+        # degree of our abilities
+        ti.mtime = committed_date
+        return ti
+    tar_args = dict(recursive=False, filter=_filter_tarinfo)
+
+    file_extension = '.{}{}'.format(
+        archivetype,
+        '{}{}'.format(
+            '.' if compression else '',
+            compression) if archivetype == 'tar' else '')
+
+    if filename is None:
+        filename = "datalad_{}".format(dataset.id)
+    if not filename.endswith(file_extension):
+        filename += file_extension
+
+    root = dataset.path
+    # use dir inside matching the output filename
+    # TODO: could be an option to the export plugin allowing empty value
+    # for no leading dir
+    leading_dir = file_basename(filename)
+
+    # workaround for inability to pass down the time stamp
+    with patch('time.time', return_value=committed_date), \
+            tarfile.open(filename, "w:{}".format(compression)) \
+            if archivetype == 'tar' \
+            else zipfile.ZipFile(
+                filename, 'w',
+                zipfile.ZIP_STORED if not compression else zipfile.ZIP_DEFLATED) \
+            as archive:
+        add_method = archive.add if archivetype == 'tar' else archive.write
+        repo_files = sorted(repo.get_indexed_files())
+        if isinstance(repo, AnnexRepo):
+            annexed = repo.is_under_annex(
+                repo_files, allow_quick=True, batch=True)
+        else:
+            annexed = [False] * len(repo_files)
```

Comment:
would be nice to have it tested on some non-annex repo just to make sure it doesn't blow somewhere with assumption of being under annex